### PR TITLE
[GSDESCARTE-9] Prints scalebar

### PR DESCRIPTION
**Pull request by tkohr**
PR allows to print a scalebar indicated in the spec. An options object permits to set its `position` and `units`.

### CODE REVIEW

File: src/printer/scalebar.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    ++i;
+  }
+  return {
+    width: width,
+    scalenumber: count,
+    suffix: suffix,
+  };
+}
+
+function createScaleBar(ctx, frameState, scaleParams, spec) {
```
Comment (fgravin) [inline src/printer/scalebar.js:113]:
```suggestion
function renderScaleBar(ctx, frameState, scaleParams, spec) {
```

File: src/printer/scalebar.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  ctx.fillText([scaleTitle], xOffset, yzero - fontsize1);
+  // Number with units
+  ctx.strokeText([scaleText], xzero + 5, yzero + fontsize1 / 2);
+  ctx.fillText([scaleText], xzero + 5, yzero + fontsize1 / 2);
+
+  // Stroke
+  ctx.beginPath();
+  ctx.lineWidth = line1 + 2;
+  ctx.strokeStyle = '#000000';
+  ctx.fillStyle = '#ffffff';
```
Comment (fgravin) [inline src/printer/scalebar.js:166]: Set 2 var on top of the Fn for odd/even colors, will be easier to change/configure later.

**Review by fgravin — CHANGES_REQUESTED** [review]
Ok excellent work, the rendering is beautiful. We will see with the final adjustements later on with the customer.
I made some comments though.
It's a bit a pity that most of the code is duplicated from OL scaleline control, I wonder if it would'nt be worth it to open an issue on openlayers to see if some refactoring would be possible to have this code exportable. I will think about this.
Thanks for your work

File: src/main/index.js [hunk 3]
```diff
@@ -33,6 +33,7 @@ export { downloadBlob } from './utils';
  * @property {[number, number]|[number, number, string]} size Width and height in pixels, or in the specified unit in 3rd place; valid units are `px`, `mm`, `cm`, `m` and `in`.
  * @property {[number, number]} center Longitude and latitude of the map center.
  * @property {number} dpi Dot-per-inch, usually 96 for a computer screen and 300 for a detailed print.
+ * @property {boolean | Object} scaleBar Indicates whether scalebar should be printed.
```
Comment (fgravin) [inline src/main/index.js:36]: `object` should be typed

File: src/printer/scalebar.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
+import { getPointResolution, METERS_PER_UNIT } from 'ol/proj';
+import ProjUnits from 'ol/proj/Units';
+import { assert } from 'ol/asserts.js';
+
+export function printScaleBar(ctx, frameState, spec) {
```
Comment (fgravin) [inline src/printer/scalebar.js:5]: All the functions of the project are typed with typescript annotation, please do it as well.
If you use an object like `scaleBar` params, create a new type for the options, thanks.

File: src/printer/scalebar.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  let scaleParams = getScaleBarParams(frameState, spec);
```
Comment (fgravin) [inline src/printer/scalebar.js:6]:
```suggestion
  const scaleParams = getScaleBarParams(frameState, spec);
```

File: src/printer/scalebar.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+      pointResolution *= 39.37;
+    } else if (nominalCount < 1609.344) {
+      suffix = 'ft';
+      pointResolution /= 0.30480061;
+    } else {
+      suffix = 'mi';
+      pointResolution /= 1609.3472;
+    }
+  } else {
+    assert(false, 33); // Invalid units
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/printer/scalebar.js]: oh yes good point, I changed it to `console.error()`

File: src/printer/scalebar.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    projection,
+    frameState.viewState.resolution,
+    center,
+    pointResolutionUnits
+  );
+
+  let nominalCount = minWidth * pointResolution;
+  let suffix = '';
+
+  if (units == Units.DEGREES) {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/printer/scalebar.js:43]: yes that's true. this comes from the updateElement() method which touches the DOM elements of the control (and needs a map). we need the same logic to use on the canvas.

File: src/printer/scalebar.js [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+export function printScaleBar(ctx, frameState, spec) {
+  let scaleParams = getScaleBarParams(frameState, spec);
+  createScaleBar(ctx, frameState, scaleParams, spec);
+}
+
+function getScaleBarParams(frameState, spec) {
+  // default values like ol.control.ScaleLine
+  const minWidth = 64;
+  const LEADING_DIGITS = [1, 2, 5];
+  const Units = {
```
[the remark this answers is not in the document]
Comment (tkohr) [reply src/printer/scalebar.js]: thanks, I must have imported it the wrong way before

File: src/main/index.js (hunk 3 above)
Comment (tkohr) [reply src/main/index.js]: I added `ScaleBar` for the spec...

File: src/printer/scalebar.js (hunk 4 above)
Comment (tkohr) [reply src/printer/scalebar.js:11]: ... and `ScaleBarParams` for the functions.

**Comment by fgravin** [comment]
> euh `Units` ?

yes but keep it like this, OL uses `Units`  for const apparently...

**Comment by tkohr** [comment]
Thanks a lot for your reviewing @fgravin. I adapted accordingly.

**Review by fgravin — APPROVED** [review]
Thanks for the corrections

**Pull request merged**